pkg/convert/profile_extra.go

```diff
@@ -0,0 +1,73 @@
+package convert
+
+// These functions are kept separately as profile.pb.go is a generated file
+
+import (
+	"sort"
+
+	"github.com/valyala/bytebufferpool"
+)
+
+func (x *Profile) Get(sampleType string, cb func(name []byte, val int)) error {
+	valueIndex := 0
+	if sampleType != "" {
+		for i, v := range x.SampleType {
+			if x.StringTable[v.Type] == sampleType {
+				valueIndex = i
+				break
+			}
+		}
+	}
+
+	b := bytebufferpool.Get()
+	for _, s := range x.Sample {
+		for i := len(s.LocationId) - 1; i >= 0; i-- {
+			name, ok := x.findFunctionName(s.LocationId[i])
+			if !ok {
+				continue
+			}
+			if b.Len() > 0 {
+				_ = b.WriteByte(';')
+			}
+			_, _ = b.WriteString(name)
+		}
+		cb(b.Bytes(), int(s.Value[valueIndex]))
+		b.Reset()
+	}
+
+	bytebufferpool.Put(b)
```

Comment:
would it be better to do `defer bytebufferpool.Put(b)` after `b := bytebufferpool.Get()`?

Reply:
It was so initially :) I was playing with benchmark and decided to not put `defer` here: although its cost is negligible, it is a hot path. From my perspective, there is no much difference, if you think it looks better with `defer`, I'll return it back.

Reply:
No, this makes sense, I was just wondering why you decided not to do defer. I think in these cases it'd be useful to add comments, e.g:
```suggestion
	// don't use defer here as it's a hot path
	bytebufferpool.Put(b)
```

---

Now that I think about this, is this a hot path though? This will run ~ 5 times every 10 seconds, right?

Reply:
Okay, I'll add a comment.

> Now that I think about this, is this a hot path though? This will run ~ 5 times every 10 seconds, right?

The function is called from `session.Snapshot()`. To my understanding it is called `profile_types` x `sample_rate` times per second (defaults to 500).

Reply:
Default sample rate is 100 Hz

And for in gospy case that API is... deceiving. the function is called Snapshot and for rbspy and pyspy profilers it means it will take a snapshot 100 times per second. In go case it almost always returns early and only does any work every 10 seconds (https://github.com/pyroscope-io/pyroscope/blob/e0f799a732de44679a9366cb7379e61f8163a08f/pkg/agent/gospy/gospy.go#L108-L115)

Reply:
Oh, right. I'm wondering why then it consumes so much CPU (https://github.com/pyroscope-io/pyroscope/issues/294#issuecomment-887180487). I'll take a closer look.

Reply:
I decided to put `defer` so as to not confuse anyone.